src/audio/volume/volume.c

```diff
@@ -712,6 +841,59 @@ static int volume_cmd(struct comp_dev *dev, int cmd, void *data,
 	}
 }
 
+/* CONFIG_IPC_MAJOR_3 */
+#elif CONFIG_IPC_MAJOR_4
+/**
+ * \brief Used to pass standard and bespoke commands (with data) to component.
+ * \param[in,out] dev Volume base component device.
+ * \param[in] cmd Command type.
+ * \param[in,out] data Control command data.
+ * \return Error code.
+ */
+static int volume_cmd(struct comp_dev *dev, int cmd, void *data,
+		      int max_data_size)
+{
+	struct ipc4_peak_volume_config *cdata = ASSUME_ALIGNED(data, 8);
+	struct vol_data *cd = comp_get_drvdata(dev);
+	uint32_t channels_count = cd->base.audio_fmt.channels_count;
+	int i;
+
+	comp_dbg(dev, "volume_cmd()");
+
+	dcache_invalidate_region(cdata, sizeof(*cdata));
+
+	/* In IPC4 driver sends volume in Q1.31 format. It is converted
+	 * into Q1.23 format to be processed by firmware.
+	 */
+	cdata->target_volume >>= 8;
```

Comment:
Round, there's the Q_SHIFT_RND() macro available, e.g. `cdata->target_volume = Q_SHIFT_RND(cdata->target_volume , 31, 23);`